hw/top_earlgrey/sw/autogen/devicetables.c

```diff
@@ -0,0 +1,2181 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+//
+// ------------------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! -------------------//
+// PLEASE DO NOT HAND-EDIT THIS FILE. IT HAS BEEN AUTO-GENERATED WITH THE FOLLOWING COMMAND:
+//
+// util/topgen.py -t hw/top_earlgrey/data/top_earlgrey.hjson \
+//                -o hw/top_earlgrey/ \
+//                --rnd_cnst_seed \
+//                1017106219537032642877583828875051302543807092889754935647094601236425074047
+
+
+
+#include "hw/top_earlgrey/sw/autogen/devicetables.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include <stdint.h>
+
+// Device tables for adc_ctrl
+_Static_assert(kDtAdcCtrlRegBlockCount == 1, "Reg block count mismatch");
+_Static_assert(kDtAdcCtrlClockCount == 2, "Clock count mismatch");
+_Static_assert(kDtAdcCtrlIrqTypeCount == 1, "IRQ count mismatch");
+_Static_assert(kDtAdcCtrlCount == 1, "Number of adc_ctrl modules mismatch");
+
+const dt_adc_ctrl_t kDtAdcCtrl[kDtAdcCtrlCount] = {
```

Comment:
Have you considered using a unified type for all the peripherals?
This type would have all possible fields, but some would be left empty when it does not make sense for a particular block.
This should be safe as we expect the tests to use the accessor functions to read from the device table, so all the checks will be done by autogenerated code.

I'm not entirely convinced this would be a batter approach, but it would reduce the amount of auto-generated code.
But it may prevent LTO to do a batter work?

Reply:
Yes, but I think it would have some drawbacks. It's still less type safe unless you had plenty of runtime checks, whereas by using a unique type, you literally can't write wrong code unless you really want to.
Also if you have some arrays (like `pins`) in the data then you would need either to point to array somewhere (more indirection and the data is not trivially copyable) or make the array of the maximal possible size.

Another advantage of using a unique type for each IP is that in the future, it is possible that some IP may want to embed more information in the DT and having a different type for each one makes that cleaner in my opinion.